[oneDAL] Data management section (#141)
oneapi-src/oneAPI-spec: source/elements/oneDAL/source/data_management/index.rst

Data management
===============

> Effective data management is among key constituents of the performance of a data

Reviewer: I'd not use 'effective', 'among key' etc. Those are just specs which describe the component of the library. Pls, use style and language of the text block removed

analytics application. For |dal_short_name|, effective data management requires
effectively performing the following operations:

- raw data acquisition
- filtering and normalization
- data conversion to a numeric representation
- data streaming to an algorithm

Key concepts
============

|dal_short_name| provides a set of customizable interfaces to operate on
out-of-memory and in-memory data in different usage scenarios, which include
:term:`batch processing <Batch mode>`, :term:`online processing <Online mode>`,
and complex scenarios.

> Table

Reviewer: The text above lists the most common ops on the data, acquisition, filtering etc (I'd also group them to highlight the stages, first bullet is about data acquisition, filtering, etc second-conversion to numeric rep, 3d - streaming to the algorithm), all of them are to be described below with respective concepts.
I'd do Key concepts first - it is okay, but I'd move def of the dataset from tables section to it
Then I'd have a section about data sources
Then a section about tables, which starts with general defs/description and introduction to concepts associated with it, etc, then (we need to analyze two options, one is flat in this version of the doc, another-have builder and accessor as subsections of the table section) 
When you discuss/define the table, describe its concept from the perspective of the problems to be solved with it:
- data storage/representation of the heterogeneous (in the general case) data, variety of the formats, dense, sparse, array of structures, etc
- table construction, different scenarios, memory ownership, interaction with data source, user's memory etc
- access to the data that returns the block in required layout  
Use still high level but precise lang. This section would help connect with next (sub)sections on the storage (we  seem to need a subsection on this), builders, accessors. Then sections about e.g. builder will give the additional info/next level (but still high) of details

-----

To represent a :term:`dataset` in a numeric :term:`data format`,
|dal_short_name| introduces a :term:`table` concept - an :term:`immutable
<Immutability>` tabular view of a :term:`dataset` where table rows represent
:term:`observations <Observation>` and columns represent :term:`features
<Feature>`.

.. image:: _static/table.png
  :width: 400
  :alt: Table

|dal_short_name| defines a set of table types, each represent data in
various :term:`data formats <Data format>`. See :ref:`tables` section for more
details about requirements of table types and table concept.

> Data source

Reviewer 2: There are multiple data sources, right?

I don't think this section provides a clear explanation of what data source is. 

This content should be more structured. 
After the definition for a data source, I think you need to have something like "data sources allow to:" and a bullet list with operations and explanations for why and when these operations might be needed

-----------

Data source defines an interface to extract data from raw sources like files,
databases, etc. and to represent them as a :term:`table` object. It can
automatically transform non-numeric :term:`categorical <Categorical feature>`
and :term:`continuous <Continuous feature>` values of data into a numeric
representation.

> It can also provide an abilities to filter or normalize :term:`feature` values
> being extracted or recover missing values:

Reviewer 2: ```suggestion
It also allows to filter and normalize :term:`feature` values
being extracted as well as recover missing values:
```

- It is not a rare situation when only a subset of features can be measured at a
  given moment. In this case, the non-measured features in the feature vector
  become blank, or missing. Special statistical techniques enable recovery
  (emulation) of missing values.

- When a given measurement introduces an :term:`outlier` or when an observed
  object can produce an :term:`outlier` by nature, special statistical techniques

>   enable detection of outliers and recovery of the abnormal data.

Reviewer 2: this list is a bit confusing in my opinion

> Data are not always fit the local memory, especially when processing with
> accelerators. Data source provides an interface to load data by batches and
> orchestrates its efficient extraction directly into local memory of the device.
> Therefore data source enables complex scenarios of data analytics, like
> :term:`online computations <Online mode>`.

Reviewer 2: ```suggestion
Data do not always fit into the local memory, especially when being processed with
accelerators. A data source provides an interface to load data by batches and
efficiently extract it directly into the device's local memory.
Therefore, a data source enables complex data analytics scenarios, such as
:term:`online computations <Online mode>`.
```

Table builder
-------------

> Some :term:`table` objects require complex procedures to create them. To make

Reviewer: The same comment about the style. Instead of setting context on 'some table requires ..' start with formal precise def which generally gives additional info/more details about builder. Explain what problem it solves using precise but still high level lang, etc. Look at the text from the perspective of the reader who sees this text for the first time and wants to get a clear understanding about the builder  after reading this section. The same comment applies to other sections

easy-of-use table creation from data of various :term:`data formats <Data
format>` or other objects of different types, |dal_short_name| introduces a
special :term:`builder` concept for table types. See :ref:`table builders
<table_builders>` section for details.

Accessor
--------

There are a lot of :term:`data formats <Data format>` used in :term:`datasets
<Dataset>` to store the data. |dal_short_name| defines different types of
:term:`tables <Table>`, table builders and other objects to support different
logical types of data and various :term:`data formats <Data format>`. However,
all these objects have different logic to access their underlying data.

Accessor allows:

- to make uniform access to the data from various sets of objects of different
  types.

- to perform efficient access to the data without exposing implementation
  details of a data container.

> - to give an ability to focus on the data acquisition in the desired :term:`data
>   format` with a wanted set of operations over the data.

Reviewer 2: ```suggestion
- to acquire data in a desired :term:`data format` for which 
a specific set of operations is defined
```

See :ref:`accessors` section for details.

Relations between concepts
==========================

> Typical workflow with data in |dal_short_name| consists of next steps:

Reviewer: it seems this this part of the info should go to the beginning of the section

1. Select an appropriate data source for extracting out-of-memory data, or
   select a table builder to manually create a table object.

2. Create :term:`table` object - a numerical representation of data, with use

>    of selected data source, table builder or table objetc constructor directly.

Reviewer: if we move it higher, then we do not give the info about builders on that level

3. Run a selected :ref:`algorithm <Algorithms>` with created table object.

4. Get algorithm :ref:`results <Result>` and access their data directly or using
   selected accessor.

.. image:: _static/typical_data_flow.png
  :width: 800
  :alt: Typical data flow

> Next sequence diagram briefly shows dependencies between table, builder, and

Reviewer: This part of the info logically comes after sections with table/builder/accessor

accessor concepts.

> Builder is responcible to create a table object, which data can be accessed

Reviewer: misprint

> either directly using table interface or with help of accessor object. Since the

Reviewer: it seems we switch to the lower level details without setting up the context. Need a higher level, but still specific def

> table object is immutable, this is not possible to change the data inside it.

Reviewer: we should mention immutability somewhere earlier when you define table

To change the data of table object, its data should be moved inside builder
object. Once it done, data inside builder can be changed by the accessor.

See sections about :ref:`tables`, :ref:`table builders <table_builders>`, and
:ref:`accessors` for details.

.. image:: _static/table_builder_accessor_sequence.png
  :width: 800

>   :alt: Sequence diagram of accessor-builder-table relations

Reviewer: Describe/discuss the content of the diagram so I'd be able to quickly understand what I have to implement for the concepts
The diagram contains colors. What do they mean
Also say about UML etc (the interactions between ... are described/represented with UML sequence diagram. It shows responsibilities, action, etc or similar

We will discuss things over the phone

Data management
===============

This section includes descriptions of concepts and objects that operate on data.
For |dal_short_name|, such set of operations, or **data management**, is
distributed between different stages of the :ref:`data analytics pipeline
<data_analytics_pipeline>`. From a perspective of data management, this pipeline
contains three main steps of data acquisition, preparation, and computation (see

> :ref:`on the picture below <typical_data_management_flow>`):

Reviewer 2: ```suggestion
:ref:`the picture below <typical_data_management_flow>`):
```

1. Raw data acquisition

  - Transfer out-of-memory data from various sources (databases, files, remote
    storages) into an in-memory representation.

2. Data preparation

  - Support different in-memory :term:`data formats <Data format>`.
  - Compress and decompress the data.
  - Convert the data into numeric representation.
  - Recover missing values.
  - Filter the data and perform data normalization.
  - Compute various statistical metrics for numerical data, such as mean, variance,

>     covariance.

Reviewer 2: ```suggestion
    and covariance.
```

3. Algorithm computation

  - Stream in-memory numerical data to the algorithm.

In complex usage scenarios, data flow goes through these three stages back and

> forth. For example, when the full data are not available at the start of the

Reviewer 2: ```suggestion
forth. For example, when data are not fully available at the start of the
```

> computation, it can be done step-by-step with blocks of data. After computation

Reviewer 2: ```suggestion
computation, it can be done step-by-step using blocks of data. After the computation
```

in-memory data during different stages of the :ref:`data analytics pipeline
<data_analytics_pipeline>`.

.. _dataset:

Dataset
--------

The main data-related concept that |dal_short_name| works with is a
:term:`dataset`. It is an in-memory or out-of-memory tabular view of data, where
table rows represent the :term:`observations <Observation>` and columns
represent the :term:`features <Feature>`.

.. image:: _static/dataset.png
  :width: 400
  :alt: Dataset

The dataset is used across all stages of the data analytics pipeline. For
example:

1. At the acquisition stage, it is downloaded into the
   local memory.

2. At the preparation stage, it is converted into a numerical
   representation.

3. At the computation stage, it is used as one of the :ref:`inputs <Input>` or
   :ref:`results <Result>` of an algorithm or a :ref:`descriptor <descriptors>`
   properties.

.. _data-source:

Data source
-----------

Data source is a concept of an out-of-memory storage for a :term:`dataset`. It is
used at the data acquisition and data preparation stages for the following:

- To extract datasets from external sources such as databases, files, remote
  storages.

> - To load datasets into the device's local memory. Data are not always fit

Reviewer 2: ```suggestion
- To load datasets into the device's local memory. Data do not always fit
```

  the local memory, especially when processing with accelerators. A data source
  provides the ability to load data by batches and extracts it directly into the
  device's local memory. Therefore, a data source enables complex data analytics
  scenarios, such as :term:`online computations <Online mode>`.

- To filter and normalize :term:`feature` values that are being extracted.

- To recover missing :term:`feature` values.

- To detect :term:`outliers <Outlier>` and recover the abnormal data.

- To transform datasets into numerical representation. Data source shall
  automatically transform non-numeric :term:`categorical <Categorical feature>`
  and :term:`continuous <Continuous feature>` data values into one of the
  numeric :term:`data formats <Data format>`.

For details, see :ref:`data-sources` section.

.. _table:

Table
-----

Table is a concept of a :term:`dataset` with in-memory numerical data. It is
used at the data preparation and data processing stages for the following:

> - To store heterogeneous in-memory data with variety of

Reviewer 2: ```suggestion
- To store heterogeneous in-memory data in a variety of
```

  :term:`data formats <Data format>`, such as dense, sparse, chunked,
  contiguous.

- To avoid unnecessary data copies during conversion from external data
  representations.

- To transfer memory ownership of the data from user application to the table,
  or share it between them.

> - To connect with the :ref:`data-source` to convert out-of-memory dataset
>   representation to in-memory view.

Reviewer 2: ```suggestion
- To connect with the :ref:`data-source` in order to convert out-of-memory dataset
  representation into in-memory view.
```

Reviewer 2: you sure it's view? not form? not dataset?

- To support streaming of the data to the algorithm.

- To access the underlying data on a device in a required :term:`data format`,

>   e.g. by blocks of the defined :term:`data layout`.

Reviewer 2: with?

For thread-safety reasons and better integration with external entities, a table
provides a read-only access to the data within it, thus, table concept
implementations shall be :term:`immutable <Immutability>`.

This concept has different logical organization and physical
:term:`format of the data <data format>`:

- Logically, a table is a :ref:`dataset` with :math:`n` rows and
  :math:`p` columns. Each row represents an :term:`observation` and each column
  is a :term:`feature` of a dataset. Physical amount of bytes needed to store

>   the data differ from the logical number of elements :math:`n \times p` within

Reviewer 2: logical number? why not just number?

For details, see :ref:`tables` section.

.. _metadata:

Metadata
--------

> Metadata concept is assotiated with a :ref:`dataset` and holds an information

Reviewer 2: ```suggestion
Metadata concept is assotiated with a :ref:`dataset` and holds information
```

about its structure and type. This information shall be enough to determine the

> particular type of a dataset, and helps to understand how to interact with

Reviewer 2: ```suggestion
particular type of a dataset, and it helps to understand how to interact with a
```

> dataset in |dal_short_name| (for example, how to use it on a particular stage of

Reviewer 2: ```suggestion
dataset in |dal_short_name| (for example, how to use it at a particular stage of
```

:ref:`data analytics pipeline <data_analytics_pipeline>` or how to access its
data).

> Metadata for all datasets shall contain:

Reviewer 2: ```suggestion
For each dataset, its metadata shall contain:
```

> - The number of rows :math:`n` and columns :math:`p` in a dataset
> 
> - Type of each :term:`feature` (e.g. :term:`nominal <Nominal feature>`,
>   :term:`interval <Interval feature>`)
> 
> - :term:`data type` of each feature
> 
> - The kind of a dataset (e.g. :ref:`table` or :ref:`data-source`)

Reviewer 2: ```suggestion
- the number of rows :math:`n` and columns :math:`p` in a dataset

- the type of each :term:`feature` (e.g. :term:`nominal <Nominal feature>`,
  :term:`interval <Interval feature>`)

- the :term:`data type` of each feature (e.g. float, double)

- the type of a dataset (e.g. :ref:`table` or :ref:`data-source`)
```

- The type of each :term:`feature` (e.g. :term:`nominal <Nominal feature>`,
  :term:`interval <Interval feature>`).

- The :term:`data type` of each feature (e.g. :code:`float` or :code:`double`).

.. note::
  Metadata can contain both compile-time and run-time information. For example,
  basic compile-time metadata is the type of a dataset - whether it is a
  particular :ref:`data-source` or a :ref:`table`. Run-time information can
  contain the :term:`feature` types and :term:`data types <Data type>` of a
  dataset.

.. _table-builder:

Table builder
-------------

> A table :term:`builder` is a concept that associated with particular

Reviewer 2: ```suggestion
A table :term:`builder` is a concept that is associated with particular
```

> :ref:`table` type and is used on data preparation and data processing stages

Reviewer 2: ```suggestion
:ref:`table` type and is used at the data preparation and data processing stages
```

  :term:`immutable <Immutability>` dataset, a builder provides the ability to
  change the values in a dataset under construction.

- Encapsulating construction process of a :ref:`table`. This is used to hide the
  implementation details as they are irrelevant for users. This also allow to
  select the most appropriate table implementation for each particular case.

- Providing additional information on how to create a :ref:`table` inside an
  algorithm for :ref:`results  <Result>`. This information includes metadata,
  memory allocators that shall be used, or even a particular table
  implementation.

For details, see :ref:`table-builders` section.

.. _accessor:

Accessor
--------

Accessor is a concept that defines a single way to get the data from an
in-memory numerical :ref:`dataset`. It allows:

- To have unified access to the data from various sets of different objects,
  such as :ref:`tables <Table>` or :ref:`table builders <table-builder>`,
  without exposing their implementation details.

- To convert a variety of numeric :term:`data formats <Data format>` into a
  smaller set of formats.

- To provide a :term:`flat <flat data>` view on the data blocks of a

>   :ref:`dataset` for better data locality. For example, some accessor

Reviewer 2: ```suggestion
  :ref:`dataset` for a better data locality. For example, some accessor
```

Data management
===============

This section includes descriptions of concepts and objects that operate with
data. For |dal_short_name|, such set of operations, or **data management**, is
distributed among different stages of a :ref:`data analytics pipeline
<data_analytics_pipeline>`. From data management angle of view, this pipeline
contains three main steps of data acquisition, preparation, and computation:

1. Raw data acquisition.

>   - Transferring out-of-memory data into in-memory representation from sources
>     like databases, files, remote storages, etc.

Reviewer 2: ```suggestion
  - Transfer out-of-memory data from various sources (databases, files, remote storages)
    into an in-memory representation.
```

2. Data preparation.

>   - Performing data normalization, filtering.
>   - Recovery of missing values
>   - Compression and decompression of the data.
>   - Conversion of data to numeric representation.
>   - Supporting different in-memory :term:`data formats <Data format>`.
>   - Computation of statistics on numerical data like mean, variance,
>     covariance, etc.

Reviewer 2: ```suggestion
  - Filter the data and perform data normalization.
  - Recover missing values.
  - Compress and decompress the data.
  - Convert the data into numeric representation.
  - Support different in-memory :term:`data formats <Data format>`.
  - Compute various statistical metrics for numerical data, including mean, variance,
    covariance, and others.
```

3. Algorithm computation.

>   - Streaming in-memory numerical data to the algorithm.

Reviewer 2: ```suggestion
  - Stream in-memory numerical data to the algorithm.
```

In complex usage scenarios, data flow can go through these three stages back and
forth. For example, when the full data are not available at the start of the
computation, it can be done step-by-step with blocks of data. After computation

> on the current block the next one should be obtained and prepared.

Reviewer 2: ```suggestion
on the current block is completed, the next block should be obtained and prepared.
```

.. image:: _static/data_management_flow.png
  :width: 800
  :alt: Typical data management flow

Key concepts
============

|dal_short_name| provides a set of concepts to operate on out-of-memory and
in-memory data during different stages of the :ref:`data analytics pipeline
<data_analytics_pipeline>`.

.. _dataset:

Dataset
--------

> The most common concept over the data in |dal_short_name| is a :term:`dataset`.

Reviewer 2: `most common` -> most widely used? the main concept?
`concept over the data` - incorrect phrase

suggestion:

```
The main data-related concept that |dal_short_name| works with is a :term:`dataset`.
```

Author: Thanks!

> This is an in-memory or out-of-memory tabular view of data, where table rows

Reviewer 2: ```suggestion
It is an in-memory or out-of-memory tabular view of data, where table rows
```

> represent :term:`observations <Observation>` and columns represent

Reviewer 2: ```suggestion
represent the :term:`observations <Observation>` and columns represent
```

> :term:`features <Feature>`.

Reviewer 2: ```suggestion
the :term:`features <Feature>`.
```

.. image:: _static/dataset.png
  :width: 400
  :alt: Dataset

> The dataset comes across all stages of the data analytics pipeline. For example:

Reviewer 2: ```suggestion
The dataset is used across all stages of the data analytics pipeline. For example:
```

> 1. On the acquisition stage, it is downloaded into the
>    local memory

Reviewer 2: ```suggestion
1. At the acquisition stage, it is downloaded into the
   local memory.
```

> 2. On the preparation stage, it is converted to numerical
>    representation

Reviewer 2: ```suggestion
2. At the preparation stage, it is converted into a numerical
   representation.
```

> 3. On the computation stage, it is used as one of the :ref:`inputs <Input>` or
>    :ref:`results <Result>` of an algorithm or even can be found in its
>    parameters

Reviewer 2: ```suggestion
3. At the computation stage, it is used as one of the :ref:`inputs <Input>` or
   :ref:`results <Result>` of an algorithm, and can even be found in its
   parameters.
```

.. _data-source:

Data source
-----------

> Data source is a concept of an out-of-memory storage of a :term:`dataset`. It is

Reviewer 2: I think it's `for` a dataset

> used on data acquisition and data preparation stages for:

Reviewer 2: ```suggestion
used at the data acquisition and data preparation stages for the following:
```

> - Dataset extraction from databases, files, remote storages, etc.

Reviewer 2: ```suggestion
- To extract datasets from databases, files, remote storages, and other sources.
```

> - Dataset loading into the local memory of the device. Data are not always fit

Reviewer 2: ```suggestion
- To load datasets into the device's local memory. Data are not always fit
```

>   the local memory, especially when processing with accelerators. Data source
>   provides an ability to load data by batches and orchestrates its efficient
>   extraction directly into local memory of the device. Therefore data source
>   enables complex scenarios of data analytics, like :term:`online computations

Reviewer 2: ```suggestion
  the local memory, especially when processing with accelerators. A data source
  provides the ability to load data by batches and orchestrates its efficient
  extraction directly into the device's local memory. Therefore, a data source
  enables complex data analytics scenarios, such as :term:`online computations
```

`orchestrates its efficient extraction directly into` -> `efficiently extract it directly into`

  <Online mode>`.

> - Filtering or normalization of :term:`feature` values being extracted.

Reviewer 2: ```suggestion
- To filter and normalize :term:`feature` values that are being extracted.
```

> - Recovery (emulation) of missing :term:`feature` values.

Reviewer 2: ```suggestion
- To recover missing :term:`feature` values.
```
do you need to mention `emulation`? i'd suggest adding the second sentence with something like this:`The recovery of missing values is also called emulation`

> - :term:`outlier` detection and recovery of the abnormal data.

Reviewer 2: ```suggestion
- To detect :term:`outliers <outlier>` and recover the abnormal data.
```

> - Dataset transformation to numerical representation. Data sources can

Reviewer 2: ```suggestion
- To transform datasets into numerical representation. Data sources can
```

  automatically transform non-numeric :term:`categorical <Categorical feature>`

>   and :term:`continuous <Continuous feature>` values of data into one of numeric

Reviewer 2: ```suggestion
  and :term:`continuous <Continuous feature>` data values into one of the numeric
```

  :term:`data formats <Data format>`.

.. _table:

Table
-----

Table is a concept of a :term:`dataset` with in-memory numerical data. It is
used on data-preparation and data-processing stages to:

- Store heterogeneous (in the general case) in-memory data with variety of
  :term:`data formats <Data format>` (dense, sparse, chunked, contiguous, etc.)

- Avoid unnecessary data copies on conversion from external data
  representations.

> - Transfer memory ownership of the data from user responsibility to the table,

Reviewer 2: from the user side to the table?

  or share it between them

> - Connect with :ref:`data-source` to convert out-of-memory dataset

Reviewer 2: with the

  representation to in-memory view

- Stream the data to the algorithm

- Access the underlying data on a device in a required :term:`data format`, e.g.

>   by blocks of defined :term:`data layout`.

Reviewer 2: of the defined

> Table provide read-only access to the data inside it for thread-safety reasons

Reviewer 2: ```suggestion
A table provides a read-only access to the data within it for thread-safety reasons
```

i'd also switch the order: `for thread-safety reasons...., a table provides ...`

> and better integration with external entities, thus, table concept is

Reviewer 2: table concept is immutable or just table?

Author: table concept requires immutability for its implementations

Reviewer 2: the table concept requires tables to be immutable -- this would be more correct than the current version

:term:`immutable <Immutability>`.

This concept may also have different logical organization and physical

> :term:`format of the data <data format>` inside it.

Reviewer 2: `inside` -> `within`
`:` at the end of the line?

> - Logically, table is a :ref:`dataset` with :math:`n` rows and

Reviewer 2: ```suggestion
- Logically, a table is a :ref:`dataset` with :math:`n` rows and
```

>   :math:`p` columns. Each row represent an :term:`observation` and each column

Reviewer 2: ```suggestion
  :math:`p` columns. Each row represents an :term:`observation` and each column
```

>   is a :term:`feature` of a dataset. Thus, table contains :math:`(n \times p)`

Reviewer 2: ```suggestion
  is a :term:`feature` of a dataset. Thus, a table contains :math:`(n \times p)`
```

  data elements, while the original (physical) amount of bytes needed to store
  the data can be different.

> - Physically, table can be organized in different ways: as a :term:`homogeneous

Reviewer 2: ```suggestion
- Physically, a table can be organized in different ways: as a :term:`homogeneous
```

>   <Homogeneous data>`, :term:`contiguous <Contiguous data>` array of bytes, as

Reviewer 2: ```suggestion
  <Homogeneous data>`, :term:`contiguous <Contiguous data>` array of bytes, as a
```

For details, see :ref:`tables` section.

.. _metadata:

Metadata
--------

Metadata concept is assotiated with a :ref:`dataset` and holds an information
about its structure and type. This information shall be enough to determine the
particular type of a dataset, and helps to understand how to interact with
dataset in |dal_short_name| (for example, how to use it on a particular stage of
:ref:`data analytics pipeline <data_analytics_pipeline>` or how to access its
data).

Metadata for all datasets shall contain:

- The number of rows :math:`n` and columns :math:`p` in a dataset

- Type of each :term:`feature` (e.g. :term:`nominal <Nominal feature>`,
  :term:`interval <Interval feature>`)

- :term:`data type` of each feature

- The kind of a dataset (e.g. :ref:`table` or :ref:`data-source`)

.. note::
  Metadata can contain both compile-time and run-time information. For example,
  basic compile-time metadata is the type of a dataset - whether it is a
  particular :ref:`data-source` or a :ref:`table`. Run-time information can
  contain the :term:`feature` types and :term:`data types <Data type>` of a
  dataset.

.. note::

>   Each concept that concretizes a dataset definition (like different tables or

Reviewer 2: `specifies` or `refines` instead of `concretizes`?
change `like` to `for example`

  data sources) shall contain additional metadata which describes this concept

>   as the whole. For example, :ref:`data-source` and :ref:`table` concepts have

Reviewer 2: ```suggestion
  as a whole. For example, :ref:`data-source` and :ref:`table` concepts have
```

>   different sets of metadata since they represent different kinds of a

Reviewer 2: `kinds` -> `types`?

  :term:`dataset`.

.. _table-builder:

Table builder
-------------

> A table :term:`builder` is a concept that associated with particular
> :ref:`table` type and is used on data-preparation and data-processing stages
> for:

Reviewer 2: ```suggestion
A table :term:`builder` is a concept that is associated with a particular
:ref:`table` type and is used at the data preparation and data processing stages
for:
```

> - Iterative (in general case) construction of a :ref:`table` from
>   another :ref:`tables <table>` or different in-memory :ref:`dataset`
>   representations

Reviewer 2: I'd remove `(in general case)`
```suggestion
- Iterative construction of a :ref:`table` from
  another :ref:`tables <table>` or a different in-memory :ref:`dataset`
  representations
```

> - Construction of a :ref:`table` from different entities that hold pieces of the

Reviewer 2: `entities that hold pieces of the dataset` seems unclear
other entities that contain information about the data? not sure it's better

>   dataset, like arrays, pointers to the memory, external entities, etc.

Reviewer 2: ```suggestion
  dataset, such as arrays, pointers to the memory, external entities, and others.
```

- Changing dataset values. Since :ref:`table` is an

>   :term:`immutable <Immutability>` dataset, builder provides an ability to
>   change the values of data in the dataset under construction.

Reviewer 2: ```suggestion
  :term:`immutable <Immutability>` dataset, a builder provides the ability to
  change the values in a dataset under construction.
```

> - Incapsulating construction process of a :ref:`table`. This is used to hide an
>   implementation details, which (1) are unnecessary for user, (2) allow to
>   select most appropriate table implementation for every particular case.

Reviewer 2: ```suggestion
- Encapsulating construction process of a :ref:`table`. This is used to hide the
  implementation details as they are irrelevant for users. This also allow to
  select the most appropriate table implementation for each particular case.
```

> - Providing an additional information how to create a :ref:`table` inside an
>   algorithm for :ref:`results  <Result>`. This information may include metadata,
>   memory allocators need to be used, or even particular implementation of a
>   table.

Reviewer 2: what is meant by `for results`? `to store the results`?
```suggestion
- Providing additional information on how to create a :ref:`table` inside an
  algorithm for :ref:`results  <Result>`. This information may include metadata,
  memory allocators that shall be used, or even a particular table implementation.
```

:ref:`table` type and is used on data-preparation and data-processing stages
for:

- Iterative (in general case) construction of a :ref:`table` from
  another :ref:`tables <table>` or different in-memory :ref:`dataset`
  representations

- Construction of a :ref:`table` from different entities that hold pieces of the
  dataset, like arrays, pointers to the memory, external entities, etc.

- Changing dataset values. Since :ref:`table` is an
  :term:`immutable <Immutability>` dataset, builder provides an ability to
  change the values of data in the dataset under construction.

- Incapsulating construction process of a :ref:`table`. This is used to hide an
  implementation details, which (1) are unnecessary for user, (2) allow to
  select most appropriate table implementation for every particular case.

- Providing an additional information how to create a :ref:`table` inside an
  algorithm for :ref:`results  <Result>`. This information may include metadata,
  memory allocators need to be used, or even particular implementation of a
  table.

For details, see :ref:`table-builders` section.

.. _accessor:

Accessor
--------

> Accessor is a concept which defines sinle way how to get the data from a
> in-memory numerical :ref:`dataset`. It allows:

Reviewer 2: ```suggestion
Accessor is a concept that defines a single way to get the data from an
in-memory numerical :ref:`dataset`. It allows:
```

> - To make uniform access to the data from various sets of different objects,
>   like :ref:`tables <Table>` or :ref:`table builders <table-builder>` without
>   exposing their implementation details.

Reviewer 2: ```suggestion
- To have unified access to the data from various sets of different objects,
  such as :ref:`tables <Table>` or :ref:`table builders <table-builder>`, without
  exposing their implementation details.
```

> - To convert a variety of numeric :term:`data formats <Data format>` into a
>   smaller set (e.g. for better vectorization).

Reviewer 2: I'd also remove `(e.g. for better vectorization)`
```suggestion
- To convert a variety of numeric :term:`data formats <Data format>` into a
  smaller set of formats (e.g. for better vectorization).
```

- To give an ability to focus on the data acquisition in the desired :term:`data

>   format` with a wanted set of operations over the data.

Reviewer 2: already commented on this: https://github.com/oneapi-src/oneAPI-spec/pull/141#discussion_r430404995

> - To make read-only, read-write and write-only access to the data.

Reviewer 2: to have? to implement?

For details, see :ref:`accessors` section.

> Use-case example for table, accessor and table builder

Reviewer 2: just `Use-case example`?

------------------------------------------------------

> This section gives an basic usage scenario of :ref:`table`, :ref:`table-builder`
> and :ref:`accessor` concepts and relations between them. The following diagram

Reviewer 2: ```suggestion
This section provides a basic usage scenario of the :ref:`table`, :ref:`table-builder`,
and :ref:`accessor` concepts and demonstrates the relations between them. The following diagram
```

shows objects of these concepts, which are highlighted by colors:

> - :ref:`table-builder` objects are blue
> 
> - :ref:`table` objects are cyan
> 
> - :ref:`accessors <Accessor>` are yellow
> 
> - objects with gray background color are not a part of |dal_short_name|
>   specification and they are provided just for illustration purposes

Reviewer 2: ```suggestion
- :ref:`Table builder <table-builder>` objects are blue.

- :ref:`Table <table>` objects are cyan.

- :ref:`accessors <Accessor>` are yellow.

- Grey objects are not a part of |dal_short_name|
  specification and they are provided just for illustration purposes.
```

> 

Reviewer 2: I think you need to insert a diagram here and then give its description

> To perform some computations based on the dataset, one shall create a
> :ref:`table` object first. It can be done using :ref:`data-source` or
> :ref:`table-builder` objects depending on the situation. The diagram briefly
> shows a situation when :ref:`table` is created from :ref:`table-builder`
> interatively with calls of various member functions that update internal state
> of the builder. Then, the `build()` method is called which takes a snapshot of
> current builder state and creates a table from it.

Reviewer 2: ```suggestion
To perform computations on a dataset, one shall create a
:ref:`table` object first. It can be done using a :ref:`data-source` or
a :ref:`table-builder` object depending on the situation. The diagram briefly
shows the situation when :ref:`table` is interatively created from a :ref:`table-builder`
with calls of various member functions that update the internal state
of the builder. Then, the `build()` method is called. It takes a snapshot of the
current builder state and creates a table from it.
```

> Once table object created, the data inside it can be accessed using member
> functions of table object or with help of read-only accessor as shown on the
> diagram. Table can be used in the computations as input or parameter of some
> algorithm.

Reviewer 2: ```suggestion
Once a table object is created, the data inside it can be accessed by its member
functions or with a help of a read-only accessor as shown on the
diagram. The table can be used as an input in computations or as a parameter of some
algorithm.
```

> Algorithms results also contain table objects. If one needs to change the data
> inside some table, builder object can be constructed for this. If table object
> is unique (just only one reference exists), it can be moved inside builder -
> thus no copy operations of data will be performed. Data inside table builder can
> be retrieved by read-only, write-only or read-write accessors.

Reviewer 2: ```suggestion
Algorithms' results also contain table objects. If one needs to change the data
within some table, a builder object can be constructed for this. If a table object
is unique (meaning that only one reference exists), it can be moved inside a builder.
Thus, no copy operations of data will be performed. Data inside a table builder can
be retrieved by read-only, write-only or read-write accessors.
```

> Accessors present on the diagram allows to get data uniformly from tables and
> table builders as blocks of rows.

Reviewer 2: ```suggestion
Accessors shown on the diagram allow to get data uniformly from tables and
table builders as blocks of rows.
```
not sure about `allow to get data uniformly`

> Internal "access_interface" object present on the diagram depicts how table and
> table builder may share one implementation with different set of operations
> defined upon it on different stages of the lifetime. It shows how accessor
> object might know how to get data from different kind of objects.

Reviewer 2: ```suggestion
Internal "access_interface" object shown on the diagram depicts how a table and
a table builder may share one implementation with a different set of operations
defined upon it at different stages of the lifetime. It shows how an accessor
object might be able to get data from different types of objects.
```

definitely don't use `know`
